pkg/storage/bloom/v1/block.go

```diff
@@ -111,3 +113,61 @@ func (bq *BlockQuerier) Err() error {
 
 	return bq.blooms.Err()
 }
+
+// CheckChunksForSeries checks if the given chunks pass a set of searches in the given bloom block.
+// It returns the list of chunks which will need to be downloaded for a query based on the initial list
+// passed as the `chks` argument. Chunks will be removed from the result set if they they are indexed in the bloom
+// and fail to pass all the searches.
+func (bq *BlockQuerier) CheckChunksForSeries(fp model.Fingerprint, chks ChunkRefs, searches [][]byte) (ChunkRefs, error) {
+	if err := bq.Seek(fp); err != nil {
+		return chks, errors.Wrapf(err, "seeking to series for fp: %v", fp)
+	}
+
+	if !bq.series.Next() {
+		return chks, nil
+	}
+
+	series := bq.series.At()
+	if series.Fingerprint != fp {
+		return chks, nil
+	}
+
+	bq.blooms.Seek(series.Offset)
+	if !bq.blooms.Next() {
+		return chks, fmt.Errorf("seeking to bloom for fp: %v", fp)
+	}
+
+	bloom := bq.blooms.At()
+
+	// First, see if the search passes the series level bloom before checking for chunks individually
+	for _, search := range searches {
+		if !bloom.sbf.Test(search) {
```

Comment:
Not sure if this hits the network, but if it does you could consider running this in parallel.

Reply:
Thanks. I've been thinking about how to control parallelism+access to the underlying data, but will do that in another PR